src/libraries/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Unix.cs

```diff
@@ -7,70 +7,135 @@
-            _namedPipeSocketHandle.DangerousAddRef(ref ignored);
-            SetHandle(_namedPipeSocketHandle.DangerousGetHandle());
+            SetPipeSocketInterlocked(namedPipeSocket, ownsHandle: true);
+            base.SetHandle(_pipeSocketHandle!.DangerousGetHandle());
         }
 
-        internal Socket? NamedPipeSocket => _namedPipeSocket;
-        internal SafeHandle? NamedPipeSocketHandle => _namedPipeSocketHandle;
+        internal Socket PipeSocket => _pipeSocket ?? CreatePipeSocket();
+
+        internal SafeHandle? PipeSocketHandle => _pipeSocketHandle;
 
         protected override void Dispose(bool disposing)
         {
             base.Dispose(disposing); // must be called before trying to Dispose the socket
-            if (disposing && _namedPipeSocket != null)
+            _disposed = 1;
+            Socket? socket;
+            if (disposing && (socket = Volatile.Read(ref _pipeSocket)) != null)
             {
-                _namedPipeSocket.Dispose();
-                _namedPipeSocket = null;
+                socket.Dispose();
+                _pipeSocket = null;
             }
         }
 
         protected override bool ReleaseHandle()
         {
             Debug.Assert(!IsInvalid);
 
-            // Clean up resources for named handles
-            if (_namedPipeSocketHandle != null)
+            if (_pipeSocketHandle != null)
             {
-                SetHandle(DefaultInvalidHandle);
-                _namedPipeSocketHandle.DangerousRelease();
-                _namedPipeSocketHandle = null;
+                base.SetHandle((IntPtr)DefaultInvalidHandle);
+                _pipeSocketHandle.DangerousRelease();
+                _pipeSocketHandle = null;
                 return true;
             }
-
-            // Clean up resources for anonymous handles
-            return (long)handle >= 0 ?
-                Interop.Sys.Close(handle) == 0 :
-                true;
+            else
+            {
+                return (long)handle >= 0 ?
+                    Interop.Sys.Close(handle) == 0 :
+                    true;
+            }
         }
 
         public override bool IsInvalid
         {
-            get { return (long)handle < 0 && _namedPipeSocket == null; }
+            get { return (long)handle < 0 && _pipeSocket == null; }
+        }
+
+        private Socket CreatePipeSocket(bool ownsHandle = true)
+        {
+            Socket? socket = null;
+            if (_disposed == 0)
+            {
+                bool refAdded = false;
+                try
+                {
+                    DangerousAddRef(ref refAdded);
+
+                    socket = SetPipeSocketInterlocked(new Socket(new SafeSocketHandle(handle, ownsHandle)), ownsHandle);
+
+                    if (_disposed == 1)
+                    {
+                        Volatile.Write(ref _pipeSocket, null);
+                        socket.Dispose();
+                        socket = null;
+                    }
```

Comment:
Help me understand this?  Why are we doing this disposed check here?

Reply:
This is meant to deal with a race that happens when `Dispose` is called concurrently with this method and `_pipeSocket` was still `null` when checked in `Dispose`.

Reply:
I see. So basically Dispose won't have nulled out the socket because it may not have been set yet, so after we create the socket, we then double check and dispose it ourselves.  Ok.  Can you add a comment?